packages/agents/src/agents/Chameleon/index.ts

```diff
@@ -86,20 +100,87 @@ export class ChameleonAgent extends Agent {
         },
         this.config.prompts.loopPreventionPrompt,
         this.config.prompts.agentSpeakPrompt,
-        this.beforeLlmResponse.bind(this)
+        this.beforeLlmResponse2.bind(this)
       );
     } catch (err) {
       this.context.logger.error(err);
       return ResultErr("Unrecoverable error encountered.");
     }
   }
 
-  protected initializeChat(args: GoalRunArgs): void {
+  protected initializeChat(): void {
     const { chat } = this.context;
-    
+
     chat.persistent(buildDirectoryPreviewMsg(this.context.workspace));
     chat.persistent("user", prompts.exhaustAllApproaches);
-    chat.persistent("user", args.goal);
+  }
+
+  protected async beforeLlmResponse2(): Promise<{ logs: ChatLogs, agentFunctions: FunctionDefinition[], allFunctions: AgentFunction<AgentContext>[]}> {
+
+    // Retrieve the last message
+    const rawChatLogs = this.context.chat.chatLogs;
+    const lastMessages = [
+      rawChatLogs.getMsg("persistent", -1),
+      rawChatLogs.getMsg("temporary", -2),
+      rawChatLogs.getMsg("temporary", -1)
+    ].filter((x) => x !== undefined) as ChatMessage[];
+
+    // Summarize large messages
+    for (let i = 0; i < lastMessages.length; ++i) {
+      const msg = lastMessages[i];
+      if (this._chunker.shouldChunk(msg)) {
+        const content = await summarizeMessage(
+          msg,
+          this.context.llm,
+          this.context.chat.tokenizer
+        );
+        lastMessages[i] = { ...msg, content };
+      }
+    }
+
+    // Predict the next step
+    const prediction = await this.askLlm(
+      new Prompt()
+        .json(lastMessages)
+        .line(`
+          Consider the above chat between a user and assistant.
+          In your expert opinion, what is the best next step for the assistant?`
+        )
+    );
+
+    const [agent, agentFunctions, persona, allFunctions] = await findBestAgent(prediction, this.context);
+
+    const context = `${persona}\n${prediction}`;
+    const maxContextTokens = this.context.llm.getMaxContextTokens();
+    const maxResponseTokens = this.context.llm.getMaxResponseTokens();
+    // TODO: remove this once we properly track function definition tokens
+    const fuzz = 500;
+    const maxChatTokens = maxContextTokens - maxResponseTokens - fuzz;
+
+    const contextualizedChat = await this._cChat.contextualize(
+      context, {
+        persistent: maxChatTokens * 0.25,
+        temporary: maxChatTokens * 0.75
```

Comment:
What's the rationale behind these factors?

Reply:
Talked about this, just a simple solution, can improve this in the future by adding a priority scalar to the chat message struct. This way we can contextualize the chat w/ a summation of `priority + relevancy` for each message